File: Go/101.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
/**
to be symmetric, for each node in the tree,
0) the 2 nodes to check can only both be nil. not symmetric is only one is nil.
1) values of left/right child nodes should equal
2) right child of left child should be symmetric to left child of right child
3) left child of right child should be symmetric to right child of left child
*/
func isSymmetric(root *TreeNode) bool {
    if root == nil {
        return true
    }
    return is(root.Left, root.Right)
}

func is(left *TreeNode, right*TreeNode) bool {
    if left==nil && right==nil {
        return true
    }
    if left==nil || right==nil {
        return false
    }
    if left.Val != right.Val {
        return false
    }
    return is(left.Left, right.Right) && is(left.Right, right.Left)
}

File: Java/101.java
/**
Recursive:
To be symmetric, below rules must hold when comparing two nodes l, r:
1) they must be both null or non-null
2) their values must be equal
3) l.left must be symmectric to r.right
4) l.right must be symmectric to r.left

Iterative:
Level-order traversal, and at each level the values must "mirror".

 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean isSymmetric(TreeNode root) {
        if(root == null) return true;
        // return check(root.left, root.right);
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root.left);
        queue.offer(root.right);

        while(!queue.isEmpty()){
            TreeNode l = queue.poll(); 
            TreeNode r = queue.poll();
            if(l == null && r == null) continue;
            if(l == null || r == null) return false;
            if(l.val != r.val) return false;
            queue.offer(l.left);
            queue.offer(r.right);
            queue.offer(l.right);
            queue.offer(r.left);
        }
        return true;
    }
    
    private boolean check(TreeNode left, TreeNode right) {
        if(left == null && right == null) return true;
        if(left == null || right == null) return false;
        if(left.val != right.val) return false;
        return check(left.left, right.right) && check(left.right, right.left);
    }
}
